packages/wallet/src/signer.ts

```diff
@@ -0,0 +1,76 @@
+import type { BytesLike } from '@ethersproject/bytes';
+import { concat, hexlify, arrayify } from '@ethersproject/bytes';
+import { ec as EC } from 'elliptic';
+
+import { hash } from './hasher';
+
+/**
+ * Return elliptic instance with curve secp256k1
+ */
+function getCurve() {
+  return new EC('secp256k1');
+}
+
+class Signer {
+  readonly address: string;
+
+  readonly publicKey: string;
+
+  readonly privateKey: string;
+
+  constructor(privateKey: BytesLike) {
+    // A lot of common tools do not prefix private keys with a 0x
+    if (typeof privateKey === 'string') {
+      if (privateKey.match(/^[0-9a-f]*$/i) && privateKey.length === 64) {
+        // eslint-disable-next-line no-param-reassign
+        privateKey = `0x${privateKey}`;
+      }
+    }
+
+    // Convert to bite array, normalize private key input allowing it to be BytesLike
+    // like remove 0x prefix and accept array of bytes
+    const privateKeyBytes = arrayify(privateKey);
+    const keyPair = getCurve().keyFromPrivate(privateKeyBytes, 'hex');
+
+    // @TODO: defineReadOnly these properties
+    // Slice(1) removes the encoding scheme from the public key
+    this.publicKey = hexlify(keyPair.getPublic(false, 'array').slice(1));
+    this.privateKey = hexlify(privateKeyBytes);
+    this.address = hash(this.publicKey);
+  }
+
+  sign(msg: BytesLike) {
```

Comment:
What do you think about renaming `sign` to `signMessage`? Message signing and transaction signing _might_ be slightly different, requiring two different implementations. At the very least, when signing an actual transaction this function will be taking a whole `Transaction` object, right? So they'll likely differ.

Reply:
I was thinking about encoding the transaction on the wallet level and sign using this method. In this way, the Signer will work for both. Maybe I could be missing some points but I can change later on the PR for the `wallet.signTransaction`.

Maybe changing the name of the param would be a good idea.

Reply:
That makes sense!